feat: Added programmable_transaction.move and related test

theiari opened the pull request:
# Description of change

Introduced `programmable_transaction.move` and `auth_context.move` in the framework, in order to represent these types in Move.

Closes #8122
Closes #8165 

## How the change has been tested

`programmable_transaction_tests` has been tested using `iota move test`

### Release Notes

theiari committed:
feat: added auth_context.move module

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+    package: ID,
+    module_name: String,
+    function: String,
+    type_arguments: vector<std::type_name::TypeName>,
```
Use `use std::type_name::TypeName;` at the top.

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+    Upgrade(UpgradeData),
+}
+
+public enum TypeInput has copy, drop {
```
This is not needed, we can use use TypeName

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+
+// === Structs ===
+
+public struct CallArg has copy, drop {
+    call_type: CallArgType,
+}
+
+public struct Command has copy, drop {
+    command_type: CommandArgType,
+}
```
No need to have a struct for these two, lets just use their Enums

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+    type_arguments: vector<std::type_name::TypeName>,
+    arguments: vector<Argument>,
+}
+
```
Could you add this as well? 
```move
public struct ProgrammableTransaction has copy, drop {
    inputs: vector<CallArg>,
    commands: vector<Command>,
}
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+
+public enum CommandArgType has copy, drop {
+    MoveCall(ProgrammableMoveCall),
+    TransferObjects(TransferObjectsData),
```
```suggestion
    TransferObjects(vector<Argument>, Argument),
```
And all the ones below as well. Only `ProgrammableMoveCall` is needed as a struct

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+
+public struct ObjectRef has copy, drop {
+    object_id: ID,
+    sequence_number: SequenceNumber,
+    object_digest: ObjectDigest,
```
```suggestion
    sequence_number: u64,
    object_digest: vector<u8>,
```
Just these are enough

miker83z commented:
No need to represent VectorType, Box and StructInput as well

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
```
```suggestion
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,290 @@
+// Modifications Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+module iota::auth_context;
```
Should be named `iota::programmable_transaction`

theiari committed:
fix: poolished auth_context.move

feat: added simple getter test

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    object_digest: vector<u8>,
+}
+
+public struct ObjectDigest has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    digest: Digest,
+}
+
+public struct SequenceNumber has copy, drop, store {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    value: u64,
+}
+
+public struct Digest has copy, drop, store {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+// Copyright (c) 2025 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+module iota::programmable_transaction;
```
Also fix the file name appropriately `programmable_transaction.move` and `programmable_transaction_tests.move`

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    MergeCoins(MergeCoinsData),
+    Publish(PublishData),
+    MakeMoveVec(MakeMoveVecData),
+    Upgrade(UpgradeData),
```
```suggestion
    SplitCoins(Argument, vector<Argument>),
    MergeCoins(Argument, vector<Argument>),
    Publish(vector<vector<u8>>, vector<ID>),
    MakeMoveVec(Option<TypeName>, vector<Argument>),
    Upgrade(vector<vector<u8>>, vector<ID>, ID, Argument),
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+
+// === Enums ===
+
+public enum CallArgType has copy, drop {
```
```suggestion
public enum CallArg has copy, drop {
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    ObjectData(ObjectArgType),
+}
+
+public enum CommandArgType has copy, drop {
```
```suggestion
public enum CommandArg has copy, drop {
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    Upgrade(UpgradeData),
+}
+
+public enum ArgumentType has copy, drop {
```
```suggestion
public enum Argument has copy, drop {
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    NestedResult(u16, u16),
+}
+
+public enum ObjectArgType has copy, drop {
```
```suggestion
public enum ObjectArg has copy, drop {
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+
+public enum ObjectArgType has copy, drop {
+    ImmOrOwnedObject(ObjectRef),
+    SharedObject(SharedObjectData),
```
```suggestion
    SharedObject {
        id: ID,
        initial_shared_version: u64,
        mutable: bool,
    },
```

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+
+// --- Command Data ---
+
+public struct TransferObjectsData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    recipient: Argument,
+}
+
+public struct SplitCoinsData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    amounts: vector<Argument>,
+}
+
+public struct MergeCoinsData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    coins_to_merge: vector<Argument>,
+}
+
+public struct PublishData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    dependencies: vector<ID>,
+}
+
+public struct MakeMoveVecData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    arguments: vector<Argument>,
+}
+
+public struct UpgradeData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    upgrade_ticket: Argument,
+}
+
+public struct Argument has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+
+// --- Helper Structs ---
+
+public struct NestedResultData has copy, drop {
```
To remove

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,260 @@
+    result_index: u16,
+}
+
+public struct SharedObjectData has copy, drop {
```
To remove

theiari committed:
refactor: remove auth_context module and related tests, migrate to programmable_transaction
fix: updated programmable_transaction_tests.move to pass both types and args for the ptb move call

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+#[test]
+fun create_ptb_move_call() {
+
+    let mut ctx = tx_context::dummy();
+    let id = object::new(&mut ctx);
+    let package_id = id.to_inner();
```
```suggestion
    let package_id = object::id_from_bytes(b"0x123");
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+    let id = object::new(&mut ctx);
+    let package_id = id.to_inner();
+
+    let mut arguments = vector::empty<Argument>();
```
```suggestion
    let mut arguments = vector[];
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+    let mut arguments = vector::empty<Argument>();
+    let input_arg = new_input_argument(0);
+
+    let mut type_names = vector::empty<TypeName>();
```
```suggestion
    let mut type_names = vector[];
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+
+    let mut type_names = vector::empty<TypeName>();
+    let tn = type_name::get<u16>();
+    vector::push_back(&mut type_names, tn);
+    vector::push_back(&mut arguments, input_arg);
```
```suggestion
    type_names.push_back(tn);
    arguments.push_back(input_arg);
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+    let call = new_move_call(
+        programmable_move_call
+    );
+    assert!(get_package_id(&programmable_move_call) == package_id);
+    assert!(get_module_name(&programmable_move_call) == b"aabb".to_ascii_string());
+    assert!(get_function_name(&programmable_move_call) == b"ccdd".to_ascii_string());
```
```suggestion
    assert!(programmable_move_call.get_package_id() == package_id);
    assert!(programmable_move_call.get_module_name() == b"aabb".to_ascii_string());
    assert!(programmable_move_call.get_function_name() == b"ccdd".to_ascii_string());
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+// SPDX-License-Identifier: Apache-2.0
+
+#[test_only]
+module iota::auth_context_tests;
```
```suggestion
module iota::programmable_transaction_tests;
```

miker83z commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Mysten Labs, Inc.
+// Modifications Copyright (c) 2024 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
```
please fix this

miker83z commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,330 @@
+
+// === CallArg Helpers ===
+
+#[test_only]
```
For providing better clarity:
- move all `[test_only]` functions at the bottom of the file, the getters should be first
- remove the `get_` part of the getters name 
- use only these sections, in this order, no subsections:
  - // === Constants ===
  - // === Structs ===
  - // === Enums ===
  - // === Public getters ===
  - // === Test-only functions ===

miker83z commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,330 @@
+// === ProgrammableMoveCall Getters ===
+
+#[test_only]
+public fun new_programmable_move_call(
```
Add an equivalent function to create a programmable transaction

theiari committed:
fix: minor changes to programmable_transaction_tests.move
refactor: changed structure of programmable_transaction.move for better clarity. removed get_ to getters
feat: added new test-only constructor for a programmable transaction

theiari committed:
feat: added auth_context.move
feat: added auth_context_tests.move

miker83z commented on crates/iota-framework/packages/iota-framework/sources/auth_context.move:
```diff
@@ -0,0 +1,60 @@
+    /// The transaction input objects or primitive values
+    tx_inputs: vector<CallArg>, //TODO: remove for 2nd iteration phase
+    /// The transaction commands to be executed sequentially.
+    tx_commands: vector<CommandArg>, //TODO: remove for 2nd iteration phase
```
LGTM! Please, rename `CommandArg` to just `Command` and then remove all TODO comments and the ones related to 2nd iteration!

theiari replied:
Do you mean 
`use iota::programmable_transaction::{CallArg, CommandArg as Command};`
OR refactor `CommandArg` enum in programmable_transaction.move to `Command`?

miker83z replied:
refactor CommandArg enum in programmable_transaction.move to Command

theiari committed:
refactor: changed enum CommandArg to Command

valeriyr commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,325 @@
+
+public fun arguments(call: &ProgrammableMoveCall): vector<Argument> {
+    call.arguments
+}
```
Let's return a vector reference in these functions to be consistent with the other ones?

TheMrAI replied:
Maybe ref to String,ID as well?

theiari replied:
Functions now return references: https://github.com/iotaledger/iota/pull/8146/commits/4a2328c37aade54e82d72da11a3daa4c922d6c78

valeriyr commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,48 @@
+    new_programmable_transaction
+};
+use std::type_name;
+
```
I think it would be nice to add the tests that check all the errors we can have, to "freeze" this behavior in the tests.
What do you think?

valeriyr reviewed (changes requested):
Looks good!
Please, check the following tests:
https://github.com/iotaledger/iota/actions/runs/17121012254/job/48562298279?pr=8146
You need to fix at least this one: `build_system_packages`.

I have a question as well. We need to update the framework snapshots after changes like this. Are we going to do this in the current PR or later?

TheMrAI commented:
> Looks good! Please, check the following tests: https://github.com/iotaledger/iota/actions/runs/17121012254/job/48562298279?pr=8146 You need to fix at least this one: `build_system_packages`.
> 
> I have a question as well. We need to update the framework snapshots after changes like this. Are we going to do this in the current PR or later?

I think we should update it here. Just to try and keep the history saner, in case we want to decide later how to merge the whole shebang.

TheMrAI commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,325 @@
+public fun pure_data(arg: &CallArg): &vector<u8> {
+    match (arg) {
+        CallArg::PureData(data) => data,
+        _ => abort EInvalidEnumVariant,
```
Shouldn't these be returning Options?
Not sure what the intent is, but `pure_data` and `object_data` seem to be fallible accessors.

theiari replied:
Regarding `pure_data` and `object_data`, I'm not entirely sure how can I return a `Option<&vector<u8>>`, because, _I think_, it's only possible without referencing. So either something like:

```
public fun pure_data(arg: &CallArg): (Option<vector<u8>>) { // no referencing
    match (arg) {
        CallArg::PureData(arg) => (option::some(*arg)),
        _ => (option::none()),
    }
}
```

Or as it is right now with referencing but without Option return:

```
public fun pure_data(arg: &CallArg): &vector<u8> {
    match (arg) {
        CallArg::PureData(data) => data,
        _ => abort EInvalidEnumVariant,
    }
}
```

TheMrAI replied:
You probably right. But we can sidestep this issue too if we simply omit it. User can match it themselves on the CallArgs (unless I am wrong again on move)

TheMrAI commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,325 @@
+    }
+}
+
+public fun command_variant_name(command: &Command): u8 {
```
Isn't the name a bit misleading?
What happens is we convert an enum into an numerical value. Which isn't a name, at least in my languages (XD).
Could we perhaps name it something like: `command_to_int` or something similar?

theiari replied:
Yeah, the name was quite nasty. I changed the function name as you suggested: https://github.com/iotaledger/iota/pull/8146/commits/4a2328c37aade54e82d72da11a3daa4c922d6c78

TheMrAI commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,325 @@
+    }
+}
+
+public fun argument_nested_result(arg: &Argument): (u16, u16) {
```
Yeah I would consider these to be all Options.
Unless by design we want to fail. But given that this is a library I think the user should be allowed to make that choice.

TheMrAI commented on crates/iota-framework/packages/iota-framework/tests/auth_context_tests.move:
```diff
@@ -0,0 +1,48 @@
+use iota::programmable_transaction::{
+    new_pure,
+    CallArg,
+    Command,
```
It is reporting to me that these are unused inputs.
Should they be removed?

TheMrAI reviewed (changes requested):
I would seriously consider removing all functions that do:
"match (arg) {
   type => internal,
   _ => Abort
}"

These functions are bloating the interface and provide little functionality. The caller of the library should be able to decide if they want to fail if something doesn't match, if they want to test it etc.
They have all the tools at their disposal as none of the enums used are private.

On a similar notion I would remove "is_pure_data", "is_object_data" stile functions as well. Let the user decide what they want to test and how. For example if they wanted to test if it is "ImmOrOwnedObject" they have to write the test themselves. There is no need to provide these.

If you do this, you can ignore my other comments regarding outputs as well.
The interface becomes much cleaner, you need to test much less.

TheMrAI commented on crates/iota-framework/packages/iota-framework/sources/programmable_transaction.move:
```diff
@@ -0,0 +1,325 @@
+
+// === Structs ===
+
+public struct ProgrammableMoveCall has copy, drop {
```
Could we add more documentation to this struct describing what it is meant to represent and for what scope?
Try and imagine a user who is not familiar with any of this and stumbles on this code. What is it supposed to be used for, where, what it represents?

theiari committed:
fix: updated published_api with new move modules
fix: set all getters to return referenced variables
fix: added few extra comments into programmable_transaction structs
refactor: renamed command_variant_name to command_to_int for better understanding

test: updated programmable_transaction_tests for type safety and abort codes check

valeriyr commented on crates/iota-framework/packages/iota-framework/tests/programmable_transaction_tests.move:
```diff
@@ -0,0 +1,98 @@
+
+    // Test that we can access individual elements through the reference
+    // assert!((*retrieved_data)[0] == 1u8);
+    // assert!((*retrieved_data)[3] == 4u8);
```
Do we need this?

theiari replied:
Huh not really actually

miker83z replied:
[96411b6](https://github.com/iotaledger/iota/pull/8146/commits/96411b63d1c92067e34dbe7156fa8c322b1244fa)

miker83z committed:
remove comments and fix spelling
